internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,236 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package delegatingresolver defines a resolver that can handle both target URI
+// and proxy address resolution, unless:
+//   - A custom dialer is set using WithContextDialer dialoption.
+//   - Proxy usage is explicitly disabled using WithNoProxy dialoption.
+//   - Client-side resolution is explicitly enforced using WithTargetResolutionEnabled.
+package delegatingresolver
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+var (
+	// HTTPSProxyFromEnvironment will be used and overwritten in the tests.
+	HTTPSProxyFromEnvironment = http.ProxyFromEnvironment
+
+	logger = grpclog.Component("delegating-resolver")
+)
+
+// delegatingResolver implements the `resolver.Resolver` interface. It uses child
+// resolvers for the target and proxy resolution. It acts as an intermediatery
+// between the child resolvers and the gRPC ClientConn.
+type delegatingResolver struct {
+	target         resolver.Target     // parsed target URI to be resolved
+	cc             resolver.ClientConn // gRPC ClientConn
+	targetResolver resolver.Resolver   // resolver for the target URI, based on its scheme
+	proxyResolver  resolver.Resolver   // resolver for the proxy URI; nil if no proxy is configured
+	targetAddrs    []resolver.Address  // resolved or unresolved target addresses, depending on proxy configuration
+	proxyAddrs     []resolver.Address  // resolved proxy addresses; empty if no proxy is configured
+	proxyURL       *url.URL            // proxy URL, derived from proxy environment and target
+	mu             sync.Mutex          // protects access to the resolver state during updates
+}
+
+func parsedURLForProxy(address string) (*url.URL, error) {
+	req := &http.Request{URL: &url.URL{Scheme: "https", Host: address}}
+	url, err := HTTPSProxyFromEnvironment(req)
+	if err != nil {
+		return nil, err
+	}
+	return url, nil
+}
+
+// New creates a new delegating resolver that is used to call the target and
```

Comment:
FWIW this package is inside internal, for our own use, so we don't need to be quite as concerned about documentation as if it was exported.

Reply:
Got it! Noted!